### PR TITLE
Add pinpoint fields to notifications

### PR DESCRIPTION
**Pull request by sastels**
# Summary | Résumé

Add a few useful fields from the Pinpoint delivery receipts to the `notification` and `notification_history` tables.
- add new fields to the database and python models
- extract the fields out of the Pinpoint delivery receipts and store in the database

## Related Issues | Cartes liées

* https://app.zenhub.com/workspaces/notify-planning-core-6411dfb7c95fb80014e0cab0/issues/gh/cds-snc/notification-planning-core/438

# Test instructions | Instructions pour tester la modification

Kinda hard to test locally :/ Figure the real test is to test in staging after we merge.

# Release Instructions | Instructions pour le déploiement

None.

# Reviewer checklist | Liste de vérification du réviseur

- [x] This PR does not break existing functionality.
- [x] This PR does not violate GCNotify's privacy policies.
- [x] This PR does not raise new security concerns. Refer to our GC Notify Risk Register document on our Google drive.
- [ ] This PR does not significantly alter performance.
- [ ] Additional required documentation resulting of these changes is covered (such as the README, setup instructions, a related ADR or the technical documentation).

> ⚠ If boxes cannot be checked off before merging the PR, they should be moved to the "Release Instructions" section with appropriate steps required to verify before release. For example, changes to celery code may require tests on staging to verify that performance has not been affected.

### CODE REVIEW

File: migrations/versions/0461_add_pinpoint_fields.py [hunk 1]
```diff
@@ -0,0 +1,43 @@
+"""
+
+Revision ID: 0461_add_pinpoint_fields
+Revises: 0460_new_service_columns
+Create Date: 2024-10-15 18:24:22.926597
+
+"""
+import sqlalchemy as sa
+from alembic import op
+from sqlalchemy.dialects import postgresql
```
[the remark this answers is not in the document]
Comment (sastels) [reply migrations/versions/0461_add_pinpoint_fields.py]: fixed!

File: app/aws/mocks.py [hunk 2]
```diff
@@ -254,15 +254,38 @@ def pinpoint_shortcode_delivered_callback(reference=None, timestamp=1467074434,
         "originationPhoneNumber": "555555",
```
[the remark this answers is not in the document]
Comment (sastels) [reply app/aws/mocks.py:254]: nope, we don't want to ever pass it.

File: app/celery/process_pinpoint_receipts_tasks.py [hunk 3]
```diff
@@ -53,6 +53,14 @@ def process_pinpoint_results(self, response):
         provider_response = receipt["messageStatusDescription"]
         isFinal = receipt["isFinal"]
 
+        # some of these fields might be missing in the receipt
+        total_message_price = receipt.get("totalMessagePrice")
```
[the remark this answers is not in the document]
Comment (sastels) [reply app/celery/process_pinpoint_receipts_tasks.py:57]: no, want it to be `None` if it's missing.

File: tests/app/celery/test_process_pinpoint_receipts_tasks.py [hunk 4]
```diff
@@ -30,13 +31,15 @@
...
-        (pinpoint_shortcode_delivered_callback, "Message has been accepted by phone carrier"),
+        (pinpoint_delivered_callback, "Message has been accepted by phone", "+13655550100"),
+        (pinpoint_shortcode_delivered_callback, "Message has been accepted by phone carrier", "555555"),
     ],
 )
-def test_process_pinpoint_results_delivered(sample_template, notify_db, notify_db_session, callback, expected_response, mocker):
+def test_process_pinpoint_results_delivered(
+    sample_template, notify_db, notify_db_session, callback, expected_response, origination_phone_number, mocker
+):
     mock_logger = mocker.patch("app.celery.process_pinpoint_receipts_tasks.current_app.logger.info")
```
[the remark this answers is not in the document]
Comment (sastels) [reply tests/app/celery/test_process_pinpoint_receipts_tasks.py:43]: It is being used later in the method...

File: tests/app/celery/test_process_pinpoint_receipts_tasks.py [hunk 5]
```diff
@@ -56,6 +59,12 @@ def test_process_pinpoint_results_delivered(sample_template, notify_db, notify_d
     assert mock_callback_task.called_once_with(get_notification_by_id(notification.id))
```
[the remark this answers is not in the document]
Comment (sastels) [reply tests/app/celery/test_process_pinpoint_receipts_tasks.py:59]: oh.... yeah fixing this led me to realize that the mock wasn't actually mocking the right thing... so good on ya, ai reviewer.

File: app/celery/process_pinpoint_receipts_tasks.py [hunk 6]
```diff
@@ -53,6 +53,14 @@ def process_pinpoint_results(self, response):
...
+        total_carrier_fee = receipt.get("totalCarrierFee")
+        iso_country_code = receipt.get("isoCountryCode")
+        carrier_name = receipt.get("carrierName")
+        message_encoding = receipt.get("messageEncoding")
+        origination_phone_number = receipt.get("originationPhoneNumber")
```
Comment (jimleroyer) [inline app/celery/process_pinpoint_receipts_tasks.py:62]: So many informative fields that we'll be able to get new insights from. 🥳

**Review by whabanks — APPROVED** [review]
LGTM! 
✅ Migration works
✅ New values are passed to the front end from API
✅ Sending and whatnot still works fine locally

File: app/dao/notifications_dao.py [hunk 7]
```diff
@@ -208,6 +219,19 @@ def _update_notification_status(notification, status, provider_response=None, bo
         notification.feedback_subtype = bounce_response.get("feedback_subtype")
         notification.ses_feedback_id = bounce_response.get("ses_feedback_id")
         notification.ses_feedback_date = bounce_response.get("ses_feedback_date")
+    if sms_total_message_price:
```
[the remark this answers is not in the document]
Comment (jimleroyer) [reply app/dao/notifications_dao.py:222]: Yeah I was wondering if we need these conditionals too, if a `notification` field is already `None` and we assign `None` again, that should be ok. But there might be a reason of doing it that way? 👀

File: migrations/versions/0462_add_pinpoint_fields.py [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+    op.add_column("notifications", sa.Column("sms_iso_country_code", sa.VARCHAR(), nullable=True))
+    op.add_column("notifications", sa.Column("sms_carrier_name", sa.VARCHAR(), nullable=True))
+    op.add_column("notifications", sa.Column("sms_message_encoding", sa.VARCHAR(), nullable=True))
+    op.add_column("notifications", sa.Column("sms_origination_phone_number", sa.VARCHAR(), nullable=True))
+    op.add_column("notification_history", sa.Column("sms_total_message_price", sa.Float(), nullable=True))
+    op.add_column("notification_history", sa.Column("sms_total_carrier_fee", sa.Float(), nullable=True))
+    op.add_column("notification_history", sa.Column("sms_iso_country_code", sa.VARCHAR(), nullable=True))
+    op.add_column("notification_history", sa.Column("sms_carrier_name", sa.VARCHAR(), nullable=True))
+    op.add_column("notification_history", sa.Column("sms_message_encoding", sa.VARCHAR(), nullable=True))
+    op.add_column("notification_history", sa.Column("sms_origination_phone_number", sa.VARCHAR(), nullable=True))
```
Comment (jimleroyer) [inline migrations/versions/0462_add_pinpoint_fields.py:27]: We should update the batch process too. I wonder if we can do it in a different PR or if that will have unintended side-effects when the archival jobs will run. Definitely something to test.

**Review by jimleroyer — COMMENTED** [review]
Looks good! I had a few questions.

File: app/celery/process_pinpoint_receipts_tasks.py [hunk 9]
```diff
@@ -88,6 +96,12 @@ def process_pinpoint_results(self, response):
             notification=notification,
```
[the remark this answers is not in the document]
Comment (sastels) [reply app/celery/process_pinpoint_receipts_tasks.py:96]: yeah, it's define above this

File: app/dao/notifications_dao.py [hunk 10]
```diff
@@ -198,7 +198,18 @@ def country_records_delivery(phone_prefix):
     return dlr and dlr.lower() == "yes"
 
 
-def _update_notification_status(notification, status, provider_response=None, bounce_response=None):
+def _update_notification_status(
```
[the remark this answers is not in the document]
Comment (sastels) [reply app/dao/notifications_dao.py:201]: I think it's fine the way it is but not an unreasonable comment.

File: app/dao/notifications_dao.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (sastels) [reply app/dao/notifications_dao.py:222]: That's a pattern we use already but yeah, I can't really see a reason why it'd make a difference. I'll clean it up as the ai suggests and verify that it's still all good.

File: app/models.py [hunk 11]
```diff
@@ -1734,6 +1734,14 @@ class Notification(BaseModel):
     ses_feedback_id = db.Column(db.String, nullable=True)
     ses_feedback_date = db.Column(db.DateTime, nullable=True)
 
+    # SMS columns
+    sms_total_message_price = db.Column(db.Numeric(), nullable=True)
+    sms_total_carrier_fee = db.Column(db.Numeric(), nullable=True)
+    sms_iso_country_code = db.Column(db.String(), nullable=True)
```
[the remark this answers is not in the document]
Comment (sastels) [reply app/models.py:1740]: hrm, a good idea.

File: migrations/versions/0462_add_pinpoint_fields.py (hunk 8 above)
Comment (sastels) [reply migrations/versions/0462_add_pinpoint_fields.py:27]: tested locally
- sent an SMS, 
- edited the database to give those fields values since we don't process the AWS receipts locally
- also set the dates for the sms in the database to be a month ago
- edited the schedule for the archiving task `delete-sms-notifications` so it would be scheduled in a couple minutes
- reran celery locally and waited for it to run
- checked the database - sms with correct fields now in history table, and removed from notifications table 🎉

**Review by jimleroyer — APPROVED** [review]
